daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala

```diff
@@ -105,17 +113,35 @@ final case class ResultNeedContract[A](
   * To resume the computation, the caller must invoke `resume` with the following argument:
   * <ul>
   * <li>`Some(package)`, if the caller can dereference `packageId` to `package`</li>
-  * <li>`None`, if the caller is unable to dereference `packageId`
+  * <li>`None`, if the caller is unable to dereference `packageId`</li>
   * </ul>
+  *
+  * It depends on the engine configuration whether the engine will validate the package provided to `resume`.
+  * If validation is switched off, it is the callers responsibility to provide a valid package corresponding to `packageId`.
   */
 final case class ResultNeedPackage[A](packageId: PackageId, resume: Option[Package] => Result[A])
     extends Result[A]
 
+/** Intermediate result indicating that the contract id corresponding to a key is required to complete the computation.
+  * To resume the computation, the caller must invoke `resume` with the following argument:
+  * <ul>
+  * <li>`Some(contractId)`, if `key` is currently assigned to `contractId`</li>
+  * <li>`None`, if `key` is unassigned</li>
+  * </ul>
+  *
+  * The caller of `resume` has to ensure that any contract id passed to `resume` has previously been associated with
+  * a contract with `key` as a key.
+  *
+  * TODO: may `contractId` refer to an archived contract?
+  * TODO: may `resume` be called with `None` even when the key is assigned?
```

Comment:
- This cannot be called for a contract that the engine has already fetched (using `ResultNeedContract`), so `contractId` cannot refer to an contract the interpretation has archive.
- The may `resume` can be called with `None` iff the key is unassigned in the ledger.

This need to be carefully reviewed for non-unique key semantics.

Reply:
> The may resume can be called with None iff the key is unassigned in the ledger.

What happens if it is called with `None` for an assigned key?

Reply:
@remyhaemmerle-da

Reply:
> This cannot be called for a contract that the engine has already fetched (using ResultNeedContract), so contractId cannot refer to an contract the interpretation has archive.

@remyhaemmerle-da I understand your answer in the context of `ResultNeedContract`. However, we are in the context of `ResultNeedKey`. 

The Engine may return a `ResultNeedKey`, if it does not know if there is an active contract having that key. What happens, if the ledger passes a contract id cid and the engine already knows that cid is archived (possibly by the current transaction). Is the behavior defined? Will it fail? Will it assume the key is unassigned? What is the responsibility of the caller?

Recall that the ledger api server uses stale data to resolve requests from the engine.

Reply:
Let me extend a bit:

1.  The engine will never ask a question about a key it already know about. For instance, an engine will never ask a question about a key of a contract it has previously created or fetched by Id.

2. The engine assumes that the returned `cid` is consistent with the key of the question, in particular  the contract associated with `cid` has actually a key equal to the key of the question.

> What happens, if the ledger passes a contract Id `cid` and the engine already knows that `cid` is archived (possibly by the current transaction)

With the above assumptions this will not happen.

> Recall that the ledger api server uses stale data to resolve requests from the engine.

The only thing that can happen is that the engine asks for a key, then ask latter for the cid that was returned. 
If the state of the ledger has changed meanwhile, the caller can simply fail with a CONTRACT_ID_NOT_FOUND. 

> What is the responsibility of the caller?
point 2. above.